### PR TITLE
Convert the Debugging article from the wiki

### PR DESCRIPTION
**Pull request by olleolleolle**
Same as #210.

# TODO to complete this article

- [ ] JVM information is quite old (Cucumber v1.2.5 and IntelliJ 2017.1) - update it to newer versions
- [ ] add some JavaScript debugging tips (note to self: see the [Cucumber-js](https://github.com/cucumber/cucumber-js) documentation)

### CODE REVIEW

**Comment by olleolleolle** [comment]
@luke-hill I'm sorry, Git behaved badly towards me, and I re-opened a PR.

File: content/cucumber/debugging.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+  wait_until_jquery_inactive
+end
+
+def wait_until_jquery_inactive
+  Capybara.using_wait_time(Capybara.default_max_wait_time) do
+    page.evaluate_script('jQuery.active').zero?
+  end
+end
+```
+By setting an environment variable, you can cause Cucumber to use various debugging tools, and you can combine them by setting multiple environment variables.
```
Comment (mlvandijk) [inline content/cucumber/debugging.md]: replace "cause" with "tell"?

File: content/cucumber/debugging.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+{{% /block %}}
+
+{{% block "java" %}}
+Here is how to debug your scenarios on the JVM, by stepping through the steps of each scenario:
+
+1. Set a conditional breakpoint in the method `cucumber.runtime.Utils#invoke`, at the line `return targetMethod.invoke(target, args)` (line 40 in `cucumber-jvm` v1.2.5) and specify the following snippet as the condition (see also [https://www.jetbrains.com/help/idea/2017.1/breakpoints-2.html](https://www.jetbrains.com/help/idea/2017.1/breakpoints-2.html) for more details on how to set one on IDEA):
```
Comment (mlvandijk) [inline content/cucumber/debugging.md]: Specify that this uses IntelliJ. Add something similar for Eclipse. (@olleolleolle if you're not familiar with Java, I can update this part myself)

File: content/cucumber/debugging.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+2. Run your [RunCukesTest](https://github.com/cucumber/cucumber-java-skeleton/blob/master/src/test/java/skeleton/RunCukesTest.java) in debug mode
+3. Assuming you haven't set any other breakpoints, the execution will stop at `Utils#invoke`
+4. Now you can either:
+   - _Step into_ to start debugging the method implementing the first step of the scenario
+   - Or _Resume_ the execution to run the current step and jump to the next one
+5. And so on..
+{{% /block %}}
+
+{{% block "javascript" %}}
+TBD.
```
Comment (mlvandijk) [inline content/cucumber/debugging.md]: Anyone available to add some JavaScript debugging tipcs? (note to self: check the Cucumber-js documentation)

**Review by mlvandijk — COMMENTED** [review]
Thanks for getting this started @olleolleolle! Let's collaborate until it's sufficient to merge. If you & @luke-hill can do the Ruby, I can do the Java, and we'll have to find someone who could add the JavaScript.

**Review by luke-hill — APPROVED** [review]

**Comment by olleolleolle** [comment]
@mlvandijk I don't know Eclipse with Java enough to create a new example for that, thanks for offering to improve that.

Should we see this as step 1, merge and continue in another debugging-with-JS-improving PR?

**Comment by mlvandijk** [comment]
@olleolleolle I also don't know Eclipse that well (haven't used it in over a year), but I'm sure we can find some link like the one to IntelliJ.

File: content/cucumber/debugging.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+  end
+end
+```
+By setting an environment variable, you can tell Cucumber to use various debugging tools, and you can combine them by setting multiple environment variables.
+{{% /block %}}
+
+{{% block "java" %}}
+In order to debug your scenarios on the JVM, you can step through the the steps of each scenario in debug mode. The example below uses IntelliJ, but Eclipse will have similar functionality.
+
+1. Set a conditional breakpoint in the method `cucumber.runtime.Utils#invoke`, at the line `return targetMethod.invoke(target, args)` (line 40 in `cucumber-jvm` v1.2.5) and specify the following snippet as the condition (see also [https://www.jetbrains.com/help/idea/2017.1/breakpoints-2.html](https://www.jetbrains.com/help/idea/2017.1/breakpoints-2.html) for more details on how to set one on IDEA):
```
Comment (mlvandijk) [inline content/cucumber/debugging.md]: Just noticed this information is quite old (Cucumber v1.2.5 and IntelliJ 2017.1) - let's update to newer versions?

**Review by mlvandijk — CHANGES_REQUESTED** [review]

**Comment by olleolleolle** [comment]
@mlvandijk You're pushing us closer to *done*!

**Comment by mlvandijk** [comment]
Almost done @olleolleolle - will squash & merge any second now ;)
And I'll create a new issue for the JavaScript debugging

File: content/cucumber/debugging.md [hunk 5]
```diff
@@ -135,14 +135,20 @@ By setting an environment variable, you can tell Cucumber to use various debuggi
...
 
-1. Set a conditional breakpoint in the method `cucumber.runtime.Utils#invoke`, at the line `return targetMethod.invoke(target, args)` (line 40 in `cucumber-jvm` v1.2.5) and specify the following snippet as the condition (see also [Breakpoints](https://www.jetbrains.com/help/idea/breakpoints.html) for more details on how to set one on IDEA):
-
-        Package pkg = target.getClass().getPackage();
-        if (pkg == null) {
-            return false;
-        }
-        return !target.getClass().getPackage().getName().startsWith("cucumber");
- 
+1. Set a conditional breakpoint on th epart of the code you want to debug.
```
Comment (xtrasimplicity) [inline content/cucumber/debugging.md]: Just noticed that `the` has been split and joined `part` (`th epart`). :)

**Pull request merged**